sql/inspect: add protected timestamp for "now" AOST case

rafiss opened the pull request:
Previously, INSPECT jobs without a historical AS OF SYSTEM TIME clause
would not create protected timestamp records, but still used an AOST
clause with the current timestamp. If span processing took a long time
(especially with BulkLowQoS admission control), garbage collection could
occur before the query completed, resulting in "batch timestamp must be
after replica GC threshold" errors.

This change adds per-span protected timestamp protection when INSPECT
uses "now" as the AOST. The implementation uses a coordinator-based
approach where:

1. When a processor starts processing a span and picks "now" as the
    timestamp, it sends a new "span started" progress message containing
    the span and timestamp via InspectProcessorProgress.

2. The coordinator's progress tracker receives this message and calls
    TryToProtectBeforeGC for the relevant tables in that span. This
    waits until 80% of the GC TTL has elapsed before creating a PTS,
    avoiding unnecessary PTS creation for quick operations.

3. When span processing completes (existing behavior), the coordinator
    cleans up the PTS for that span. Any remaining PTS records are
    cleaned up when the tracker terminates (e.g., on job cancellation).

This coordinator-based design keeps PTS management centralized rather
than distributed across processors, simplifying cleanup and error
handling. PTS failures are logged but don't fail the job since the
protection is best-effort.

###  sql/inspect: use minimum timestamp for PTS protection

Previously, the INSPECT job called TryToProtectBeforeGC per span with
different timestamps. Since the job only stores one PTS record, each
new span's call to Protect would update the existing record's timestamp
via UpdateTimestamp, which removes protection for older spans.

To address this, this patch changes the PTS strategy to track the
minimum (oldest) timestamp across all active spans and protect only at
that timestamp. Since PROTECT_AFTER mode protects all data at or after
the specified timestamp, protecting at the minimum covers all active
spans. When the oldest span completes, the PTS is updated to the new
minimum timestamp, allowing GC of data between the old and new minimum.

Resolves: #159866
Epic: None

Release note: None

rafiss committed:
sql/inspect: add per-span protected timestamp for "now" AOST case

Previously, INSPECT jobs without a historical AS OF SYSTEM TIME clause
would not create protected timestamp records, but still used an AOST
clause with the current timestamp. If span processing took a long time
(especially with BulkLowQoS admission control), garbage collection could
occur before the query completed, resulting in "batch timestamp must be
after replica GC threshold" errors.

This change adds per-span protected timestamp protection when INSPECT
uses "now" as the AOST. The implementation uses a coordinator-based
approach where:

1. When a processor starts processing a span and picks "now" as the
    timestamp, it sends a new "span started" progress message containing
    the span and timestamp via InspectProcessorProgress.

2. The coordinator's progress tracker receives this message and calls
    TryToProtectBeforeGC for the relevant tables in that span. This
    waits until 80% of the GC TTL has elapsed before creating a PTS,
    avoiding unnecessary PTS creation for quick operations.

3. When span processing completes (existing behavior), the coordinator
    cleans up the PTS for that span. Any remaining PTS records are
    cleaned up when the tracker terminates (e.g., on job cancellation).

This coordinator-based design keeps PTS management centralized rather
than distributed across processors, simplifying cleanup and error
handling. PTS failures are logged but don't fail the job since the
protection is best-effort.

Resolves: #159866

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/160138)
<!-- Reviewable:end -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> @fqazi reviewed 3 files and all commit messages, and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160138#-:-OhHkwm9ETjg0KO77qOD:b-uwkowd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen).



<!-- Sent from Reviewable.io -->

rafiss committed:
inspect: log progress periodically

In addition to checkpointing in the job, now we also log progress to
text logs periodically in order to enhance observability.

Release note: None

rafiss committed:
jobsprotectedts: simplify TryToProtectBeforeGC to accept descpb.ID

Previously, TryToProtectBeforeGC accepted a catalog.TableDescriptor
parameter but only used it to call GetID() in two places. This was
unnecessarily restrictive and forced callers to load a full table
descriptor just to pass the ID.

This change simplifies the function signature to accept a descpb.ID
directly. The most significant improvement is in inspect/progress.go,
where this eliminates an unnecessary DescsTxn call that was only used
to load the descriptor for its ID.

Release note: None
Epic: None

rafiss committed:
sql/inspect: use minimum timestamp for PTS protection

Previously, the INSPECT job called TryToProtectBeforeGC per span with
different timestamps. Since the job only stores one PTS record, each
new span's call to Protect would update the existing record's timestamp
via UpdateTimestamp, which removes protection for older spans.

To address this, this patch changes the PTS strategy to track the
minimum (oldest) timestamp across all active spans and protect only at
that timestamp. Since PROTECT_AFTER mode protects all data at or after
the specified timestamp, protecting at the minimum covers all active
spans. When the oldest span completes, the PTS is updated to the new
minimum timestamp, allowing GC of data between the old and new minimum.

Release note: None

rafiss commented:
> doesn't this clobber the PTS ID stored in the job each time it is run? 

Thanks for pointing that out. I didn't realize that the PTS system would clobber. I had thought it would protect all the timestamps until each one had its `cleaner` called, but on further reflection the way that it works now makes sense.

I have reworked this (currently in a separate commit, but I can squash if that's preferred) so that INSPECT only protects the minimum timestamp that is currently being used by all processors. When that timestamp is done, that PTS record will be cleaned up and the coordinator finds the next smallest timestamp that needs to be protected.

> Could we go even further and use table ID in the message from the workers to the coordinator?

I don't think so. We still need to know when a specific span starts being processed and is done being processed, since the timestamp that needs to be protected then later cleaned up is associated with that specific span.

spilchen reviewed (approved):
Looks great. Just a couple of minor nits.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @spilchen partially reviewed 7 files and made 3 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160138#-:-OiIaeht4tqkBMQ1-1Xa:bpaup23)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @fqazi, @kev-cao, and @rafiss).

___
*[`pkg/sql/inspect/progress.go` line 342 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160138#-OiIWvWg-RkocAIwoxnH:-OiIWvWh9h64MTEcUdKk:b-opk56h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c344d072fa273299dcd255975adb93601d10d897/pkg/sql/inspect/progress.go#L342)):*
> ```Go
> 	if !needsNewPTS {
> 		log.VEventf(ctx, 2, "INSPECT: span %s at %s covered by existing PTS at %s",
> 			spanStarted, tsToProtect, t.mu.currentPTSTimestamp)
> ```

do we have a mutex held for `t.mu.currentPTSTimestamp`?
___
*[`pkg/sql/inspect/progress.go` line 361 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160138#-OiIY-ZqE_eosOWUNXNV:-OiIY-ZqE_eosOWUNXNW:b-5yqzbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c344d072fa273299dcd255975adb93601d10d897/pkg/sql/inspect/progress.go#L361)):*
> ```Go
> 		_, tableID, err := t.codec.DecodeTablePrefix(spanStarted.Key)
> 		if err != nil {
> 			log.Dev.Warningf(ctx, "failed to decode table ID from span %s: %v", spanStarted, err)
> ```

do we need to remove the span from `activeSpanTimestamps` if we get in here? Or can we extract the tableID sooner, right at the very start?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> @rafiss made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/160138)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @fqazi, @kev-cao, and @spilchen).

___
*[`pkg/sql/inspect/progress.go` line 342 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160138#-OiIWvWg-RkocAIwoxnH:-OiKHTWN4zWVM2O2VIcW:b-kj6lou) ([raw file](https://github.com/cockroachdb/cockroach/blob/c344d072fa273299dcd255975adb93601d10d897/pkg/sql/inspect/progress.go#L342)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

do we have a mutex held for `t.mu.currentPTSTimestamp`?
</blockquote></details>

nice catch; fixed
___
*[`pkg/sql/inspect/progress.go` line 361 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/160138#-OiIY-ZqE_eosOWUNXNV:-OiKGJnYBPCMa4aDSgUQ:b-ehf8pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c344d072fa273299dcd255975adb93601d10d897/pkg/sql/inspect/progress.go#L361)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

do we need to remove the span from `activeSpanTimestamps` if we get in here? Or can we extract the tableID sooner, right at the very start?
</blockquote></details>

i don't think we need to remove it. keeping it in `activeSpanTimestamps` actually seems desirable, since the span is still getting actively processed, even if we fail to add the protected timestamp (or fail to decode the key)


<!-- Sent from Reviewable.io -->

rafiss commented:
tftr!

bors r+
